src/main.cpp

```diff
@@ -6835,37 +6836,44 @@ bool static ProcessMessage(const CChainParams& chainparams, CNode* pfrom, string
         // Reject incoming connections from nodes that don't know about the current epoch
         const Consensus::Params& consensusParams = chainparams.GetConsensus();
         auto currentEpoch = CurrentEpoch(GetHeight(), consensusParams);
-        if (pfrom->nVersion < consensusParams.vUpgrades[currentEpoch].nProtocolVersion &&
+        if (nVersion < consensusParams.vUpgrades[currentEpoch].nProtocolVersion &&
             !(
                 chainparams.NetworkIDString() == "regtest" &&
                 !GetBoolArg("-nurejectoldversions", DEFAULT_NU_REJECT_OLD_VERSIONS)
             )
         ) {
-            LogPrintf("peer=%d using obsolete version %i; disconnecting\n", pfrom->id, pfrom->nVersion);
+            LogPrintf("peer=%d using obsolete version %i; disconnecting\n", pfrom->id, nVersion);
             pfrom->PushMessage("reject", strCommand, REJECT_OBSOLETE,
                             strprintf("Version must be %d or greater",
                             consensusParams.vUpgrades[currentEpoch].nProtocolVersion));
             pfrom->fDisconnect = true;
             return false;
         }
 
-        if (!vRecv.empty())
+        // We've successfully parsed the mandatory fields and checked the version.
+        // It's safe to leave these set even if subsequent parsing fails.
+        pfrom->nVersion = nVersion;
+        pfrom->nServices = nServices;
+
+        if (!vRecv.empty()) {
```

Comment:
Should these check the length remaining to ensure sufficient bytes remain for successful parsing? Or is the approach here to rely upon exception handling?

Reply:
The length of the message is allowed to be either the minimum length (just the mandatory fields), or exactly one of the lengths obtained by adding each optional set of fields in turn, or anything longer than the longest such length (to allow for future expansion). That's exactly what the current code allows, and yes, we're relying on exception handling to reject anything else.